File: src/Client.Web/src/media/motionDecoder.ts
import type { Fetcher } from './fetcher'
import { DecodeController, type ChunkDecoder } from './decodeController'

export interface MotionFrame {
  timestamp: number
  cells: Uint8Array
  cols: number
  rows: number
  sync: boolean
}

const headerSize = 22
const maxChains = 64

export class MotionDecoder implements ChunkDecoder<MotionFrame> {
  onFrame: ((frame: MotionFrame, gopTimestamp?: number) => void) | null = null

  private readonly controller: DecodeController<MotionFrame>
  private readonly chains = new Map<number, Uint8Array>()
  private queue: Promise<void> = Promise.resolve()

  constructor(fetcher: Fetcher) {
    this.controller = new DecodeController(fetcher, this)
  }

  setTarget(gopTimestamps: number[]) {
    this.controller.setTarget(gopTimestamps)
  }

  getFrame(ts: number): MotionFrame | null {
    return this.controller.getFrame(ts)
  }

  stats(): { gops: number, frames: number } {
    return this.controller.stats()
  }

  flush() {
    this.controller.clear()
    this.chains.clear()
    this.queue = Promise.resolve()
  }

  decode(data: Uint8Array, gopTimestamp: number) {
    this.queue = this.queue
      .then(() => this.decodeChunk(data, gopTimestamp))
      .catch(e => console.error('motion decode failed', e))
  }

  dispose(_frame: MotionFrame) {}

  private async decodeChunk(data: Uint8Array, gopTimestamp: number) {
    let offset = 0
    while (offset < data.length) {
      if (data.length - offset < headerSize) {
        console.warn('motion gop has', data.length - offset, 'trailing bytes')
        return
      }
      if (data[offset] !== 0x4D || data[offset + 1] !== 0x47
        || data[offset + 2] !== 0x52 || data[offset + 3] !== 0x44) {
        console.warn('motion gop has no MGRD magic at offset', offset)
        return
      }
      const view = new DataView(data.buffer, data.byteOffset + offset)
      const flags = data[offset + 5]
      const timestamp = Number(view.getBigUint64(6, true))
      const cols = view.getUint16(14, true)
      const rows = view.getUint16(16, true)
      const payloadLength = view.getUint32(18, true)
      if (data.length - offset < headerSize + payloadLength) {
        console.warn('motion gop truncated', data.length - offset, 'of', headerSize + payloadLength, 'bytes')
        return
      }

      const cells = await inflate(data.slice(offset + headerSize, offset + headerSize + payloadLength))
      offset += headerSize + payloadLength
      if (cells.length !== cols * rows) {
        console.warn('motion cell count mismatch', cells.length, 'for', `${cols}x${rows}`)
        continue
      }

      const sync = (flags & 0x01) !== 0
      const base = sync ? null : this.chains.get(gopTimestamp)
      for (let i = 0; i < cells.length; i++)
        cells[i] ^= base ? base[i] : 0
      this.setChain(gopTimestamp, cells)

      this.onFrame?.({ timestamp, cells, cols, rows, sync }, gopTimestamp)
    }
  }

  private setChain(gopTimestamp: number, cells: Uint8Array) {
    this.chains.delete(gopTimestamp)
    this.chains.set(gopTimestamp, cells)
    while (this.chains.size > maxChains)
      this.chains.delete(this.chains.keys().next().value!)
  }
}

async function inflate(data: Uint8Array): Promise<Uint8Array> {
  const ds = new DecompressionStream('deflate-raw')
  const writer = ds.writable.getWriter()
  const reader = ds.readable.getReader()
  writer.write(data as Uint8Array<ArrayBuffer>)
  writer.close()
  const chunks: Uint8Array[] = []
  while (true) {
    const { done, value } = await reader.read()
    if (done) break
    chunks.push(value)
  }
  const total = chunks.reduce((n, c) => n + c.length, 0)
  const out = new Uint8Array(total)
  let offset = 0
  for (const chunk of chunks) { out.set(chunk, offset); offset += chunk.length }
  return out
}

File: src/Client.Core/Decoding/MotionDecoder.cs
using System.Buffers.Binary;
using System.IO.Compression;
using Microsoft.Extensions.Logging;

namespace Client.Core.Decoding;

public sealed class MotionDecoder : IChunkDecoder<MotionFrame>
{
  private const int HeaderSize = 22;
  private const int MaxChains = 64;

  private readonly ILogger _logger;
  private readonly DecodeController<MotionFrame> _controller;
  private readonly Dictionary<ulong, byte[]> _chains = [];
  private readonly List<ulong> _chainOrder = [];

  public MotionDecoder(ILogger logger, Fetcher fetcher)
  {
    _logger = logger;
    _controller = new DecodeController<MotionFrame>(fetcher, this);
  }

  public void SetTarget(ulong[] gopTimestamps) => _controller.SetTarget(gopTimestamps);

  public MotionFrame? GetFrame(long ts) => _controller.GetFrame(ts);

  public (int Gops, int Frames) Stats() => _controller.Stats();

  public void Flush()
  {
    _controller.Clear();
    _chains.Clear();
    _chainOrder.Clear();
  }

  public void Decode(ReadOnlyMemory<byte> data, ulong gopTimestamp)
  {
    var offset = 0;
    while (offset < data.Length)
    {
      var unit = data.Span[offset..];
      if (unit.Length < HeaderSize)
      {
        _logger.LogWarning("Motion gop has {Bytes} trailing bytes", unit.Length);
        return;
      }
      if (unit[0] != 'M' || unit[1] != 'G' || unit[2] != 'R' || unit[3] != 'D')
      {
        _logger.LogWarning("Motion gop has no MGRD magic at offset {Offset}", offset);
        return;
      }

      var flags = unit[5];
      var timestamp = (long)BinaryPrimitives.ReadUInt64LittleEndian(unit[6..]);
      var cols = BinaryPrimitives.ReadUInt16LittleEndian(unit[14..]);
      var rows = BinaryPrimitives.ReadUInt16LittleEndian(unit[16..]);
      var payloadLength = (int)BinaryPrimitives.ReadUInt32LittleEndian(unit[18..]);
      if (unit.Length < HeaderSize + payloadLength)
      {
        _logger.LogWarning("Motion gop truncated: {Have} of {Need} bytes",
          unit.Length, HeaderSize + payloadLength);
        return;
      }

      var cells = Inflate(data.Slice(offset + HeaderSize, payloadLength));
      offset += HeaderSize + payloadLength;
      if (cells.Length != cols * rows)
      {
        _logger.LogWarning("Motion cell count mismatch: {Count} for {Cols}x{Rows}",
          cells.Length, cols, rows);
        continue;
      }

      var sync = (flags & 0x01) != 0;
      if (!sync && _chains.TryGetValue(gopTimestamp, out var baseFrame))
        for (var i = 0; i < cells.Length; i++)
          cells[i] ^= baseFrame[i];
      SetChain(gopTimestamp, cells);

      _controller.PushFrame(new MotionFrame(timestamp, cells, cols, rows, sync), gopTimestamp);
    }
  }

  public void Dispose(MotionFrame frame) { }

  private void SetChain(ulong gopTimestamp, byte[] cells)
  {
    _chainOrder.Remove(gopTimestamp);
    _chainOrder.Add(gopTimestamp);
    _chains[gopTimestamp] = cells;
    while (_chains.Count > MaxChains)
    {
      _chains.Remove(_chainOrder[0]);
      _chainOrder.RemoveAt(0);
    }
  }

  private static byte[] Inflate(ReadOnlyMemory<byte> compressed)
  {
    using var input = new MemoryStream(compressed.ToArray());
    using var deflate = new DeflateStream(input, CompressionMode.Decompress);
    using var output = new MemoryStream();
    deflate.CopyTo(output);
    return output.ToArray();
  }
}
